src/libraries/System.Private.CoreLib/src/System/IO/Strategies/AsyncWindowsFileStreamStrategy.cs

```diff
@@ -35,27 +35,10 @@ private unsafe ValueTask<int> ReadAsyncInternal(Memory<byte> destination, Cancel
                 ThrowHelper.ThrowNotSupportedException_UnreadableStream();
             }
 
-            long positionBefore = _filePosition;
-            if (CanSeek)
-            {
-                long len = Length;
-                if (positionBefore + destination.Length > len)
-                {
-                    destination = positionBefore <= len ?
-                        destination.Slice(0, (int)(len - positionBefore)) :
-                        default;
-                }
-
-                // When using overlapped IO, the OS is not supposed to
-                // touch the file pointer location at all.  We will adjust it
-                // ourselves, but only in memory. This isn't threadsafe.
-                _filePosition += destination.Length;
-            }
```

Comment:
This means that any Windows code being ported from .NET Framework to .NET 6 that was issuing concurrent reads based on position will now be broken.  Just want to confirm that's intentional.

Reply:
We have already announced that for `FileStream` with buffering enabled, which is the default setting so I think it's OK: https://github.com/dotnet/runtime/issues/50858

Reply:
The other option is to enable consistency in the other direction: always update the position before the operation.  Net result of this would be that on Unix we'd need to also patch the position back after a read if fewer bytes were read.  But if we're saying that you can't trust the position until after the operation completes anyway, maybe that's ok and a smaller break?

Reply:
> The other option is to enable consistency in the other direction: always update the position before the operation

On Unix, it would require us to perform a sys-call to obtain the file length (which we should not cache on Unix, because file locking is only advisory and others can still write to file) to make sure that the position is not `>= EOF`. If we would set `_position > EOF`, the next write operation could create a sparse file. Sample:

```cs
File.WriteAllBytes(path, new byte[100]);
using var fs = new FileStream(path, FileMode.Open, FileAccess.ReadWrite, FileShare.ReadWrite, bufferSize: 0, FileOptions.Asynchronous);
fs.ReadAsync(new byte[123]); // no await, the position is 123
fs.WriteAsync(new byte[] { 1 }); // the byte would get written at position 123, not 100
```

Reply:
@stephentoub you are right that this could break users who issue concurrent reads in .NET Framework. With https://github.com/dotnet/runtime/issues/50858 we break them only in a way that the reads are not concurrent, they are serialized. With this change they could perform multuple reads at the same offset without knowing that. I am going to close the PR

Reply:
> it would require us to perform a sys-call to obtain the file length (which we should not cache on Unix, because file locking is only advisory and others can still write to file) to make sure that the position is not >= EOF

That's why I said "Net result of this would be that on Unix we'd need to also patch the position back after a read if fewer bytes were read"

> If we would set _position > EOF, the next write operation could create a sparse file

As it currently stands, if you issue a write concurrently with a read, you'll be issuing it to the non-updated position and you'll corrupt the file, anyway.  How is creating a sparse file worse?

Reply:
> With this change they could perform multuple reads at the same offset without knowing that. I am going to close the PR

👍

Reply:
(I still wonder if we should change the Unix behavior, though, to first update the position and then patch it back after the read completes.  It would seem to enable more compatibility with existing Windows code, even though with the Unix implementation either way you shouldn't be relying on the Position while there's an active async operation in flight.)

Reply:
I really don't like the fact that we used to pretend that issuing concurrent reads and writes is OK with non-thread-safe `FileStream`. In a perfect world, we would always update the position after the operation completes and give compiler errors for invalid usage.

But getting back to your question I think that we can only choose the lesser evil here:

* If we update the position after read completes, we have great perf (single sys-call), but the users who are misusing `FileStream` in the following way:
    * issuing multiple concurrent reads when buffering is disabled and the reads are not awaited: they end up reading content from the same offset (silent error)
    * issuing write before async read operation completes: they end up overwriting the file content (silent error)
* If we update the position before read starts using current file length:
    * everyone has to pay for the perf penalty (additional sys-call)
    * if someone has extended the file in the meantime we still need to patch the position. Perhaps we could use some opportunistic caching anyway? 
* If we update the position before the read starts assuming that we are going to read entire buffer (`_position += buffer.Length`):
    * there is no perf penalty (we are not checking the file length up-front)
    * issuing concurrent reads should be fine, as long as `pread` returns `0` when reading from `offset > EOF`
    * issuing write before async read operation completes: we might end up with a sparse file. This should be extremely rare, as this is extreme misuse.

@stephentoub which option do you like the most? I am going to verify the `pread` result for `offset > EOF` in the meantime

Reply:
You probably guess given my previous comments, but I think option 3 (update position before read starts and then backpatch it if necessary) is the least bad.

Reply:
> I really don't like the fact that we used to pretend that issuing concurrent reads and writes is OK with non-thread-safe FileStream

100% agree. If we were doing it from scratch today, we would not do that.